src/poetry/puzzle/provider.py

```diff
@@ -478,15 +516,25 @@ def complete_package(self, package: DependencyPackage) -> DependencyPackage:
             "url",
             "git",
         }:
-            package = DependencyPackage(
-                package.dependency,
-                self._pool.package(
-                    package.name,
-                    package.version.text,
-                    extras=list(package.dependency.extras),
-                    repository=package.dependency.source_name,
-                ),
-            )
+            try:
+                package = DependencyPackage(
+                    package.dependency,
+                    self._pool.package(
+                        package.name,
+                        package.version.text,
+                        extras=list(package.dependency.extras),
+                        repository=package.dependency.source_name,
+                    ),
+                )
+            except PackageNotFound as e:
+                for pkg in self.search_for_installed_packages(package.dependency):
+                    package = DependencyPackage(
+                        package.dependency,
+                        pkg,
+                    )
+                    break
```

Comment:
I'm not quite sure what this break is for -- if we just intend to get the first package in the list, a call to `next()` might be more explicit.

Reply:
More to trigger the else block in the for loop, we can also do `next` and then catch the `StopIteration`. For me this is a bit cleaner; but I am happy to switch to `next` too.

Reply:
Ah, I missed the for-else -- I think I would prefer a next/StopIteration as the pattern is less likely to get lost in indentation noise (like just happened to me) and the intent is more explicit.

Reply:
Fine! Here I was ... all excited I could use a `for/else` :D